Mining Mini-Rework (formerly Golem enhancements)

VmpOS2NHSkhkejA9 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
psa: if I say difficulty or level I mean seismic level.

This PR primarily reworks golems, making them more unique, making their AI far more lethal and adjusting spawns to complement both a more controlled learning curve for new players and making higher difficulties signifigantly harder. It also contains some rebalancing of ore yields and tweaks to mining equipment.

---
<details>
<summary>Improve, replace and add golem types:</summary>

### Improve, replace and add golem types:

#### Level 1:
Coal and iron golems. They're mostly unchanged from before.

#### Level 2:
Silver golems: ranged enemies that fire stun bolts.

#### Level 3:
Platinum golems: A melee enemy that charges at players for a high damage attack unless dodged.
Uranium golems: Mostly unchanged from before. They heal nearby golems as before, but now they will also try to follow targets at a distance without themselves attacking.

#### Level 4:
Coal golems are replaced by graphite golems: They grab players instead of dealing damage normally, locking them in place for other golems to deal a killing blow.
Plasma golems: Suicide bombers. They can no longer be defused, and will begin detonating on their own if they enter melee range.

#### Level 5:
Gold golems: Ranged lethal golems, they spawn spikes under their target, which will damage non-golems standing on them if they don't move out of the way quickly enough.
Silver golems replaced by ancient silver golems: They retreat if their target gets too close, making them harder to kill with a single melee attack.
###### gold golems are a reference to something...
#### Level 6:
Diamond golems: Unchanged. Very high health tanks.
Ansible golems: Mostly the same as before, except they'll maintain a distance away from their target, fleeing if too close. They also have a 10% chance to drop a bluespace crystal.

### Rework golem spawns to be weighted and follow a few difficulty-dependent rules:
- Spawns are weighted based on the difficulty, with higher difficulties introducing different types and replacing easier enemy types with tougher variants of themselves, ie, coal golem -> graphite golem at difficulty 4
- Golems spawn in groups of 3, 4 or 5 depending on the difficulty

### Signifigant golem AI improvements:
- Golems have had their view range increased by 1, making it signifigantly more difficult to pick them off from the edge of your view. Barring diagonals: if you can see them, they can see you.
- Golems will share their target with their group - if one spots you, all of the ones nearby will. You cannot fight golems one-by-one.
</details>

---

<details>
<summary>Other changes</summary>

#### Ore Distribution Rebalancing

Ore veins have been made fewer, larger and they scale less with seismic level. Previously, the rarity of more valuable materials, vein count AND ore per tile scaled with seismic level, making low seismic levels disproportionately poor and high ones excessively rich. Now, only ore per tile scales with seismic level. Oh, and surface veins were severely nerfed. You _will_ interact with this system if you want an abundance of ore. Finally, hydrogen is now a level 6 exclusive: Everything except cave generation treats hydrogen as the rarest material, and now cave generation does too.

#### Equipment Adjustments
- The KARL can now be recharged by 25% with a golem core.
- Ore satchels now hold 200 ore, instead of 100.
- Miners no longer get a shotgun or jackhammer. They are relics from before the KARL. They still get a pickaxe, if they for some reason need it.


</details>

---

I hope this PR can see some player testing before a full merge; it's not really possible for me to balance higher difficulties alone when they're designed to be impossible alone.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Mining is boring, repetitive, horrifically unbalanced and frankly trivial. The highest difficulty is a cakewalk and they're effectively an infinite money machine for the guild the moment they turn up if they have any idea what they're doing.

This PR, hopefully, changes all of the above. By varying the enemies and cranking up the difficulty, mining can be made more engaging and we can lock abundances of every material you could ever dream of behind teamwork and preparation.

There are also a mountain of bugfixes in this PR. I'd love to list them all in the CL, but I can't remember what bugs I introduced and what bugs were already there before I fixed them. I can't exactly list a bugfix for a bug that only ever existed on my local.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Testing
Individually tested golems on the ship by spawning them and testing their behaviour. All of them work individually in a vacuum, and the AI improvements work with groups. 
I confirmed golems spawn properly in caves, in the groups they are expected to spawn in with proper spacing between spawns.
I tried to test them in the wild, by playing some miner and doing a seismic level 2 cave. This failed, miserably. I cannot test gameplay currently because it's too hard for me.

https://github.com/user-attachments/assets/a445f52f-21d2-4d77-a842-21b762256f6f

(please ignore the bgm I don't know how to edit it out)

<!-- Describe the tests you ran with your addition. It is recommended to add images, videos and step-by-step explanations of conducted testing. -->

## Changelog
:cl:
add: Gold golems; lethal ranged enemies that summon spikes under their target
balance: massive difficulty increase for cave mining; Spawns have been reworked to spawn golems in groups with types dependent on seismic level. Additionally, new types have been added, some have been reworked and their AI has been tuned up to let them see further and attack in coordinated groups, making it far more difficult to pick them off alone.
del: Miner's shotgun and jackhammer.
imageadd: added icons for gold golem spikes
code: Basic grab functionality for superior animals; their AI still isn't set to do this, so it must be trigged externally by each mob that intends to grab.
del: Tile-based resources. Nothing actually used them, anyway.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

VmpOS2NHSkhkejA9 commented:
there is also a handful of code changes affecting grabs and superior animals that I needed to make to make some of this possible (mostly the enhanced coal golems honestly) so I'm praying for no merge conflicts

VmpOS2NHSkhkejA9 commented:
This is in a functional state now, and works as intended. The balance in seismic levels 4, 5 and 6 are untested; I can't test a difficulty alone if that difficulty is designed to be near impossible. Hopefully this can end up on the test server for a bit before merge.

TheShown911 commented on code/datums/movement/mob.dm:
```diff
@@ -249,7 +249,7 @@
 		return MOVEMENT_STOP
 
-	for(var/obj/item/grab/G in mob.grabbed_by)
+	for(var/G in mob.grabbed_by)
```
You can probably remove this entire loop

VmpOS2NHSkhkejA9 replied:
This loop is a core part of grabs? removing it would severely break grabs. People would just be able to walk out of all grab types without using the resist button at all.

TheShown911 replied:
No? It just returns MOVEMENT_STOP from the proc. There is no need for it to be a loop or iterate over anything. You can just make a normal if statement to check `mob.grabbed_by` length and if it is (not 0), just return immediately.

VmpOS2NHSkhkejA9 replied:
oh, I thought you meant remove it entirely, not replace it with something else.

TheShown911 commented on code/modules/mining/drilling/cave_generator.dm:
```diff
@@ -588,16 +595,86 @@
+
+				var/list/mob/living/carbon/superior_animal/golem/golems_to_spawn = list()
+
+				switch(seismic_lvl)
```
I think it would be simpler and easier to balance to have a dictionary of golem weights for each seismic level instead of this messy switch with a lot of ternary statements but i guess it works

VmpOS2NHSkhkejA9 replied:
see [difficulties.dm](https://github.com/discordia-space/CEV-Eris/pull/8553/files#diff-16ebb0f2d584a15176b3462d414859f818027c86d576a2e91427566c62c16d95)

TheShown911 commented on code/modules/mob/living/carbon/superior_animal/golem/golem.dm:
```diff
@@ -21,19 +21,9 @@
-
-// OneStar patrol borg that defends OneStar facilities
+GLOBAL_LIST_EMPTY(all_golems) // golems check this list to loop over allies
+GLOBAL_LIST_EMPTY(active_golems) // smaller list that only contains golems with a current target mob
```
We can make a golem SS for this :mitosis:

TheShown911 reviewed (approved):
Probably good, should be tested

VmpOS2NHSkhkejA9 commented:
I've made a bunch more changes. Honestly, I'm not sure I can list them accurately, but I _will_ try:
<details>
<summary>Changelog?</summary>

- Rewrote golem spawn logic, spawn locations are now more consistent, and spawn pools are defined in datums (see difficulties.dm).
- Replaced golem ore types with ore datums, hopefully making them more consistent with other ore sources. Additionally, golem ore drops scale with seismic level.
- Reduced ore-per-tile's scaling with seismic level.
- Made golem cores recharge the KARL by 25%
- Gave ansible golems a 10% chance to drop a single bluespace crystal.
- Increased ore bag storage to 200 ore.
- Slightly lowered the layer of ore, so they will not bury anything on the same tile,
- Reduced the frequency of ore veins, but increased their size.
- Made caves guarantee at least one vein of each type.
- Examining the deep drill now gives you a rough estimate of how many ore tiles are still in a cave.
- Made the deep drill's health work on switches instead of a lengthy if statement (why did I do it that way initially?) and removed that one pointless ternary operator.
- Made hydrogen a seismic level 6 exclusive.
- Reduced ore vein count on the asteroid surface.
- Removed jackhammer, toolmods, shotgun, shotgun shells from mining lockers.

</details

VmpOS2NHSkhkejA9 commented:
not sure why tests are failing. ceil and floor are built in byond procs, and it compiles + runs fine on my machine.

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/golem/golem.dm:
```diff
@@ -106,3 +115,89 @@ GLOBAL_LIST_INIT(golems_special, list(/mob/living/carbon/superior_animal/golem/s
 		if(obstacle)
 			obstacle.attack_generic(src, rand(surrounds_mult * melee_damage_lower, surrounds_mult * melee_damage_upper), pick(attacktext), TRUE)
+
+/mob/living/carbon/superior_animal/golem/loseTarget()
+	GLOB.active_golems -= src
+	. = ..()
```
I don't seem to find a moment when anything is added to `active_golems` list, let alone when golem acquires a new target.

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/golem/golem.dm:
```diff
@@ -106,3 +115,89 @@ GLOBAL_LIST_INIT(golems_special, list(/mob/living/carbon/superior_animal/golem/s
+			target_mob = findTarget()
+			if(target_mob)
+				stance = HOSTILE_STANCE_ATTACK
+				for(var/mob/living/carbon/superior_animal/golem/ally in GLOB.all_golems)
```
![maneuver-1825954183](https://github.com/user-attachments/assets/1149b4a0-51a0-4fd0-8403-d14140870b36)

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/golem/types/gold.dm:
```diff
@@ -0,0 +1,86 @@
+		QDEL_IN(src, 20)
+
+
+
+
+
```
```suggestion
#undef GOLD_SPIKE_COOLDOWN
#undef GOLD_SPIKE_WINDUP
#undef GOLD_SPIKE_DAMAGE
```
Since these defines aren't used outside of this file, there is no need for preprocessor to keep track of them past this point.

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/golem/types/plasma.dm:
```diff
@@ -31,49 +30,30 @@
-// Special capacity of plasma golem: blow up upon death except if hit with melee weapon
 /mob/living/carbon/superior_animal/golem/plasma/death(gibbed, message = deathmessage)
 	if(det_status == DET_STABLE)
+		src.Stun(10) //prevent any movement. not actually sure if this is necessary for a dead mob.
```
```suggestion
		Stun(10) // Prevents movement. Not actually sure if this is necessary for a dead mob.
```
It isn't. But if you want to keep this anyway, then there is no need for `src`.

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/golem/types/plasma.dm:
```diff
@@ -31,49 +30,30 @@
-		. = ..()
+			// Plasma ball on location
+			visible_message(SPAN_DANGER("\The [src] explodes into a ball of burning palsma!"))
+			for(var/turf/floor/target_tile in range(2, loc))
```
```suggestion
			for(var/turf/floor/target_tile as anything in RANGE_TURFS(2, loc))
```
Macro is a lot faster and returns a list of turf references instead of _literally everything_ in range, so `as anything` skips unnecessary type checks.

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/golem/types/plasma.dm:
```diff
@@ -31,49 +30,30 @@
+			visible_message(SPAN_DANGER("\The [src] explodes into a ball of burning palsma!"))
+			for(var/turf/floor/target_tile in range(2, loc))
+				new /obj/effect/decal/cleanable/liquid_fuel(target_tile, 2, 1)
+				spawn (0) target_tile.hotspot_expose((T20C * 2) + 380, 500)  // From flamethrower code
```
Firstly, it should be
```suggestion
				spawn(0)
					target_tile.hotspot_expose((T20C * 2) + 380, 500)  // From flamethrower code
```
Secondly, `spawn()` and `sleep()` do some weird scheduling behind the scene that is an absolute cancer performance-wise and issue becomes multiplicatively worse with number of calls.
So unless you _absolutely must_ to use them - don't.
If you do, I wanna know what goes wrong without `spawn()` and what workarounds you have tried.

VmpOS2NHSkhkejA9 replied:
Removed `spawn()`, works without issues. Also fixed the spelling issues, erroneous death message and fixed them being able to move when detonating.

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/golem/types/plasma.dm:
```diff
@@ -31,49 +30,30 @@
+
+/mob/living/carbon/superior_animal/golem/plasma/UnarmedAttack(atom/A, proximity)
+	if(det_status == DET_STABLE)
+		src.death(FALSE, FALSE)
```
```suggestion
		death(FALSE, FALSE)
```

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/attack.dm:
```diff
@@ -1,11 +1,16 @@
 	return
 
-/mob/living/carbon/superior_animal/UnarmedAttack(atom/A, var/proximity)
+/mob/living/carbon/superior_animal/UnarmedAttack(atom/A, var/proximity, var/attackdamage)
```
```suggestion
/mob/living/carbon/superior_animal/UnarmedAttack(atom/A, proximity, attackdamage)
```

SirRichardFrancis commented on code/modules/mining/drilling/difficulties.dm:
```diff
@@ -0,0 +1,151 @@
+/datum/cave_difficulty_level/proc/get_golem_spawns()
+	var/list/golems_to_spawn = list() //list of golem types to spawn
+	if(golem_count_melee)
+		for(var/c = 0, c < golem_count_melee, c++)
+			golems_to_spawn += pickweight(golem_weights_melee)
```
`for(var/c in 0 to golem_count_melee)` style is noticeably more performant, or so other coders say.
Probably some kind of optimization on BYOND side.
Or rather lack of optimization when C-style loops are being converted from DM.

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/golem/types/uranium.dm:
```diff
@@ -42,10 +42,11 @@
-			G.adjustFireLoss(-GOLEM_REGENERATION) // Burn Regeneration
 	. = ..()
+	if(target_mob)
+		for(var/mob/living/carbon/superior_animal/golem/ally in GLOB.active_golems)
```
https://github.com/user-attachments/assets/4debcca8-d59e-4c9c-95ba-f44c723617f6

VmpOS2NHSkhkejA9 replied:
I'm not sure I understand, I figured this would be a decent way to go about this?

SirRichardFrancis replied:
It _is_ a decent way actually.
No ideal, but unless we have a whole lot of uranium golems running around it shouldn't noticeably lag.
For optimization ideas look at SSchinks. Having a cooldown for this ability may help too.

VmpOS2NHSkhkejA9 replied:
Having this ability run any less often could start to impact gameplay. If it happens too infrequently players would be able to deal damage to golems _inbetween_ healing cycles, essentially bypassing the uranium golem's effect.

VmpOS2NHSkhkejA9 replied:
Also, is there a way to measure the performance of different procs? I'm not actually entirely convinced `getMobsInRangeChunked()` would be faster, since it would loop over a lot more mobs than the _at most_ 10-15 and usual of around 5 that this would be checking over per uranium golem.

SirRichardFrancis replied:
There is built-in profiler (see BYOND documentation) and I've heard some coders successfully used [tracy](https://github.com/wolfpld/tracy) for this, I personally haven't.
But most commonly it would be done by calling a proc bajillion times and simply measuring time it took to for the proc to run, comparing with other procs and implementations as needed.

VmpOS2NHSkhkejA9 replied:
Tested and `getMobsInRangeChunked()` is around 10x faster than the method I used and thought would be cleaner. Swapped all golem code to use it instead.

SirRichardFrancis commented on code/modules/mob/living/carbon/superior_animal/superior_animal.dm:
```diff
@@ -374,3 +388,27 @@
+	breakgrab()
+	. = ..()
+
+/mob/living/carbon/superior_animal/proc/simplegrab(var/mob/living/target) // superior animals won't do this naturally, but this proc makes it easy to implement such behaviour in specific mobs
```
```suggestion
/mob/living/carbon/superior_animal/proc/simplegrab(mob/living/target) // superior animals won't do this naturally, but this proc makes it easy to implement such behaviour in specific mobs
```

SirRichardFrancis commented:
> not sure why tests are failing. ceil and floor are built in byond procs, and it compiles + runs fine on my machine.

Mentioned this in review.
TL;DR: Codebase uses 514 version of BYOND that doesn't support procs from newer versions.

VmpOS2NHSkhkejA9 commented on code/game/objects/items/weapons/tools/karl.dm:
```diff
@@ -134,27 +137,40 @@
 	set src in view(1)
 
+	if(usr.incapacitated() || !Adjacent(usr))
+		to_chat(usr, "<span class='warning'>You can't do that.</span>")
```
of note I copied this from modpc code so that code also needs to be changed in the future if this is how this should be done

VmpOS2NHSkhkejA9 commented on code/modules/mining/drilling/cave_mineral.dm:
```diff
@@ -35,8 +35,7 @@
 	var/seismic_multiplier = 1
-
-	has_resources = 1
+	var/obj/cave_generator/cave_gen
```
cringe? howso? I'm not sure how to do this better. I'll add the de-reference in `Destroy()` but if there's a better way to track the number of veins still in a cave I'd love to hear it.

SirRichardFrancis replied:
I recall there being a limit of one cave existing at a time, is that still true?
If so, there is no need for any object references, let alone on every cave turf, ore count could be in some variable instead.
Like in a subsystem, `SSmapping.cave_ore_count` or some such. If there are multiple caves, then this variable could be a list with each cave/cave_gen having an ID that acts as index to the list.

VmpOS2NHSkhkejA9 replied:
To clear things up this is not every cave turf, only every vein turf. Non-vein cave walls are `/turf/mineral_unsimulated`, and asteroid floors are `/turf/floor/asteroid`. 

There is still a limit of one cave, so I'll add this as a var to SSmapping instead.

VmpOS2NHSkhkejA9 commented:
all requested changes have been made and I did some extremely brief testing without finding issues.
still not sure why linters are failing.

VmpOS2NHSkhkejA9 commented on code/modules/mob/living/carbon/superior_animal/golem/golem.dm:
```diff
@@ -74,20 +67,36 @@ GLOBAL_LIST_INIT(golems_special, list(/mob/living/carbon/superior_animal/golem/s
+/mob/living/carbon/superior_animal/golem/Destroy()
+	GLOB.all_golems -= src
+	GLOB.active_golems -= src
+	..()
```
done

VmpOS2NHSkhkejA9 commented:
Removed the golem vars from SSmobs and changed golems to use `getMobsInRangeChunked()`. That should be every requested change handled now.

hyperioo commented on code/modules/mining/drilling/cave_generator.dm:
```diff
@@ -461,23 +503,21 @@
+											/datum/cave_vein/uranium = CAVE_VWEIGHT,
+											/datum/cave_vein/diamond = CAVE_VWEIGHT * 0.75,
+											/datum/cave_vein/platinum = CAVE_VWEIGHT * 0.75,
+											/datum/cave_vein/hydrogen = CAVE_VWEIGHT * ((seismic_lvl == 6) ? (CAVE_VWEIGHT * 0.5) : 0))//only generate on max difficulty. super duper valuable.
```
Is the CAVE_VWEIGHT^2 intended? Or should it just be `CAVE_VWEIGHT * 0.5`

VmpOS2NHSkhkejA9 replied:
oops
fixed

hyperioo commented on code/modules/mob/living/carbon/superior_animal/golem/types/ansible.dm:
```diff
@@ -53,17 +51,18 @@
+/mob/living/carbon/superior_animal/golem/ansible/death()
+	if(prob(10)) //10% chance to drop a bs crystal. since ansible golems are only present on max difficulty this seems like a fair reward.
+		var/obj/item/bluespace_crystal/crystal = new /obj/item/bluespace_crystal(loc)
+		visible_message(SPAN_NOTICE("A [crystal.name] falls out of [src] as it disintegrates."), 1)
```
```suggestion
		visible_message(SPAN_NOTICE("A [crystal.name] falls out of [src] as it disintegrates."))
```
In case the golem is killed by a ranged attacker, the message would be missed.

VmpOS2NHSkhkejA9 replied:
done

hyperioo commented on code/modules/mob/living/carbon/superior_animal/golem/types/coal.dm:
```diff
@@ -27,14 +27,18 @@
+
+/mob/living/carbon/superior_animal/golem/coal/enhanced/UnarmedAttack(atom/A, proximity)
+	if(istype(A, /mob/living/carbon))
+		visible_message(SPAN_DANGER("<b>[src]</b> grabs at [target_mob]!"), 1)
```
Could be useful to let other people further than 1 tile away that the victim is being grabbed (after all they're supposed to see it even if they're not right next to the golem)

hyperioo commented on code/modules/mob/living/carbon/superior_animal/golem/types/platinum.dm:
```diff
@@ -27,15 +33,65 @@
+		charge_cooldown = world.time + PLATINUM_CHARGE_CD
+
+		walk(src,0) // halt movement
+		visible_message(SPAN_DANGER("<b>[src]</b> stops and prepares to charge at [target_mob]!"), 1)
```
All these short-range `, 1)` are making me doubt myself. The proc is `/atom/proc/visible_message(message, blind_message, range = world.view)` right?

VmpOS2NHSkhkejA9 replied:
I think I just copied these from somewhere else and put my own text in to be honest. This just seems like an oversight on my part.

hyperioo commented on code/modules/mob/living/carbon/superior_animal/golem/types/platinum.dm:
```diff
@@ -27,15 +33,65 @@
+						victim.explosion_act(PLATINUM_CHARGE_DAMAGE_OBSTACLES * 4) //TEAR THROUGH ALL THAT IMPEDES YOU
+					break // if the turf is blocked (ie a wall/door/window), stop charging here
+
+			visible_message(SPAN_DANGER("<b>[src]</b> [pick(charge_verbs)] at [target_mob]!"), 1)
```
Same?

hyperioo commented on code/modules/mob/living/carbon/superior_animal/superior_animal.dm:
```diff
@@ -374,3 +388,27 @@
+	else if(!target || !Adjacent(target))
+		return
+
+	visible_message("<span class='warning'>[src] grabs [target]!</span>")
```
```suggestion
	visible_message(SPAN_WARNING("[src] grabs [target]!"))
```

hyperioo reviewed (commented):
Good looking overall! Great job!👍

VmpOS2NHSkhkejA9 commented on code/modules/mob/living/carbon/superior_animal/golem/golem.dm:
```diff
@@ -74,20 +64,35 @@ GLOBAL_LIST_INIT(golems_special, list(/mob/living/carbon/superior_animal/golem/s
+/mob/living/carbon/superior_animal/golem/Initialize(var/mapload, difficulty)
+	if(mineral_name && (mineral_name in ore_data))
+		mineral = ore_data[mineral_name]
+	difficultylevel = difficulty
```
difficultylevel is a datum so defaulting to 1 would be a bad idea

VmpOS2NHSkhkejA9 replied:
Adjusted the way golem_ore_mult is passed onto golems, golems no longer store a reference to the datum and ore mult defaults to one.
